pkg/controller/node/node_controller.go

```diff
@@ -0,0 +1,251 @@
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNode) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// log.Printf("Reconciling Node %s/%s\n", request.Namespace, request.Name)
+
+	reconcilePeriod := 5 * time.Second
+	reconcileResult := reconcile.Result{RequeueAfter: reconcilePeriod}
+
+	// Fetch the Node instance
+	instance := &corev1.Node{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcileResult, err
+	}
+
+	// Get the current storageos cluster.
+	cluster, err := r.findCurrentCluster()
+	if err != nil {
+		if err == ErrCurrentClusterNotFound {
+			// Do not requeue request. No StorageOS cluster is installed.
+			return reconcile.Result{}, nil
+		}
+		// Requeue the request in order to retry getting the cluster.
+		log.Println("failed to find current cluster:", err)
+		return reconcileResult, err
+	}
+
+	// Set a storageos cluster client.
+	if r.stosClient != nil {
+		// Compare the cluster names, generations and UUIDs to check if it's
+		// the same cluster. Update the client if client cluster name,
+		// generation or UID are different from current cluster.
+		if r.stosClient.clusterName != cluster.GetName() ||
+			r.stosClient.clusterGeneration != cluster.GetGeneration() ||
+			r.stosClient.clusterUID != cluster.GetUID() {
+			r.setClientForCluster(cluster)
+		}
+	} else {
+		// No previous client. Create a new client.
+		r.setClientForCluster(cluster)
+	}
+
+	// Sync labels to StorageOS node object.
+	if err = r.syncLabels(instance.Name, instance.Labels); err != nil {
+		log.Println("failed to sync labels:", err)
+		// Error syncing labels - requeue the request.
+		return reconcileResult, err
+	}
+
+	return reconcileResult, nil
+}
+
+// SyncNodeLabels applies the Kubernetes node labels to StorageOS node objects.
+func (r *ReconcileNode) syncLabels(name string, labels map[string]string) error {
+	if len(name) == 0 || len(labels) == 0 {
+		return nil
+	}
+
+	// Get StorageOS node
+	node, err := r.stosClient.Node(name)
+	if err != nil {
+		return err
+	}
+
+	// Initialize the map if empty.
+	if len(node.Labels) == 0 {
+		node.Labels = make(map[string]string)
+	}
+
+	// Check if the k8s node labels already exist in storageos node labels.
+	// If there's no update, or addition, do not update the storageos node.
+	changed := false
+	for k, v := range labels {
+		// If the label already exists, compare the new and old values.
+		if v2, ok := node.Labels[k]; ok {
```

Comment:
Just checking, as this is additive a label will never be removed right? 

I am a little worried about:

* Add label `replicas=3`
* Change your mind
* Remove label

I imagine we don't want to remove labels set that aren't StorageOS labels? Perhaps we can do some sort of set diff and remove entries in `node.Labels` that are not in `labels` that are prefixed by `storageos.com` only.

Reply:
Yes, we never remove labels.
AFAIK, we don't have any internally used labels for nodes. There are volume and pool labels https://docs.storageos.com/docs/reference/labels.
In order to avoid any stale node labels in the future, there had been discussions to restrict who can modify node labels, like allowing only the operator to modify node labels. But for now, we are only adding the labels and updating any existing label with new values.

Reply:
We're double-checking what the delete behaviour should be.  The user-agent has been set, so we could decide to only allow the operator to modify node labels.

As it is currently, if you want to delete a node label that was inherited from k8s, you need to remove it from k8s and then remove it from the StorageOS node.  I think that's ok for initial version.

Reply:
As briefly touched on in the CP meeting earlier it should be possible to do something like:

```go
// Global state in the operator
var (
	// prevNodeLabels maps node names to a set of previously observed kubernetes
	// labels for that node.
	//
	// It should only be accessed when holding the prevNodeLabelsMu lock (this
	// should probably be encapsulated into a type to only allow locked access).
	prevNodeLabels   = map[string]map[string]string{}
	prevNodeLabelsMu = &sync.Mutex{}
)

func syncLabels(name string, etc ...) {
	// Grab the old labels, and compare them to the new labels.
	//
	// This diff only operates on the k8s labels, so custom StorageOS labels
	// applied directly (not through k8s) remain unchanged.
	prevNodeLabelsMu.Lock()
	oldLabels := prevNodeLabels[name]
	prevNodeLabelsMu.Unlock()

	changes := diff(newLabels, oldLabels)

	// Apply changes
}
```

This lets the k8s labels be freely changed by the user of k8s, including removing them when they're removed in k8s, and it doesn't touch any labels set via StorageOS directly. 

Obviously it would have to be reset when the cluster changes, and it wouldn't work for cold-starts - in that case we would have to fall back to the additive-only state (or we could read the current labels at operator startup if we want to give better guarantees?).

All of this only works if the calls to `Reconcile` are linearised/single-threaded per node, and `oldLabels` is obviously a map that needs it's own lock if there can be concurrent calls to syncLabels for a given node - perhaps a struct that combines a lock and the map would be best as the value of the `prevNodeLabels` map.

Realise this might be not desired right now, but would be a nice usability addition to keep in mind!

Reply:
Discussed with @darkowlzz - we think this will work well and we'd like to add, but in a separate PR.  Planning to do it next week.